### PR TITLE
ZCS-15980: updated to process %z in timezone data

### PR DESCRIPTION
**Pull request by k-kato**
**Background:**
* `timezones.ics` for Zimbra server is created on [zm-timezones](https://github.com/Zimbra/zm-timezones). `zmtzdata2ical` executes `zmjava com.zimbra.common.calendar.ZoneInfo2iCalendar`
* `tzdata2024b` provided by iana (https://www.iana.org/time-zones) introduced `%z` for time zone abbreviations. (See NEWS in tzdata2024b.tar.gz)

Now `TZNAME` in `timezones.ics` needs to be generated by replacing `%z` with an appropriate value. Otherwise, existing TZNAME are removed from the ics.

**Fix:**
Update `ZoneInfo2iCalendar.java` to replace `%z`. The value is generated using timezone offsets.
TZNAME will be set like this.
```
BEGIN:STANDARD
TZNAME:+0330   // HERE
DTSTART:16010101T000000
TZOFFSETTO:+0330
TZOFFSETFROM:+0330
END:STANDARD
END:VTIMEZONE
BEGIN:VTIMEZONE


BEGIN:STANDARD
TZNAME:-03/-02  // HERE
DTSTART:16010101T020000
TZOFFSETTO:-0300
TZOFFSETFROM:-0200
RRULE:FREQ=YEARLY;WKST=MO;INTERVAL=1;BYMONTH=11;BYDAY=1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:-03/-02   // HERE
DTSTART:16010101T020000
TZOFFSETTO:-0200
TZOFFSETFROM:-0300
RRULE:FREQ=YEARLY;WKST=MO;INTERVAL=1;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
```
FYI, to and from offsets are based on standard and dailight offsets, and they are opposite in standard and daylight saving time
```
new lines 378-384
        if (isStandard) {
            toOffset = standardOffset;
            fromOffset = daylightOffset;
        } else {
            toOffset = daylightOffset;
            fromOffset = standardOffset;
        }
```
but a value of `%z` can be generated using `standard/daylightOffset`, not `to/fromOffset`. It keeps compatibility as before.

**Related PR:**
* https://github.com/Zimbra/zm-timezones/pull/37

### CODE REVIEW

**Review by silentsakky — APPROVED** [review]

**Review by lavegupta — APPROVED** [review]

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java [hunk 1]
```diff
@@ -324,6 +324,9 @@ private static Observance toObservanceComp(int hintYear, RuleLine rline, boolean
         PropertyList props = new PropertyList();
         String tzname = getObservanceName(tznameFormat, rline);
         if (tzname != null) {
+            if (tzname.equals("%z")) {
```
Comment (shrutig0510) [inline common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: Please create constant for "%z"

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java [hunk 2]
```diff
@@ -397,8 +400,31 @@ private static Observance toObservanceComp(int hintYear, RuleLine rline, boolean
         }
     }
 
+    private static String replaceTzNameVariable(Time standardOffset, Time daylightOffset) {
+        String stdOffset = null;
+        String dlOffset = null;
+        stdOffset = new TzOffsetTo(new UtcOffset(getUtcOffset(standardOffset))).getValue();
+        stdOffset = stdOffset.replaceAll("00$", "");
```
Comment (shrutig0510) [inline common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: Please create constant for "00$"

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java [hunk 3]
```diff
@@ -397,8 +400,31 @@ private static Observance toObservanceComp(int hintYear, RuleLine rline, boolean
...
+            return stdOffset;
+        } else {
+            return stdOffset + "/" + dlOffset;
+        }
+    }
+
     private static Standard toStandardComp(Time gmtOffset, String tznameFormat) {
         PropertyList props = new PropertyList();
+        if (tznameFormat != null && tznameFormat.equals("%z")) {
+            if (tznameFormat.equals("%z")) {
```
Comment (shrutig0510) [inline common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: Do we require this if. Above if statement already handle this!

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java (hunk 1 above)
Comment (k-kato) [reply common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: added `Z_PARAM`

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java (hunk 2 above)
Comment (k-kato) [reply common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: added `ZERO_MINUTES_REGEX`

File: common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java (hunk 3 above)
Comment (k-kato) [reply common/src/java/com/zimbra/common/calendar/ZoneInfo2iCalendar.java]: You're right. Thank you for pointing it out.

**Review by shrutig0510 — APPROVED** [review]

**Pull request merged**